### DISCUSSION TITLE
Tensor contractions

### DISCUSSION THREAD

**Discussion by matthiasneuner** [idaholab/moose#18194] [Show and tell]
Hi,
I just saw pull request  #18174  which introduces new higher order tensor contractions.

 I thought it would be useful to promote the existence of Fastor (https://github.com/romeric/Fastor), which is a templated, header only, vectorized Tensor library aiming at similar functionality as MOOSE's 2nd, 3rd and 4th order tensor implementations.

However, its mighty Einstein summation functionality allows to write tensor operations in a  convenient and readable way.
I make use of Fastor in a lot of projects (including MOOSE applications, e.g.  https://github.com/matthiasneuner/chamois/blob/08ce1819ef6931539b9c2e3a6d1bfa4f32d95e8c/src/materials/ComputeMarmotMaterialGradientEnhancedMicropolar.C#L222)

Best regards, Matthias

**Comment by hugary1995** [comment]
Could you leave a comment in that PR? I'm very supportive of adding an Einstein summation method. Manually adding all the contractions wouldn't be tractable in the long run and is very difficult to maintain.

**Comment by dschwen** [comment]
Thanks Matthias, this is a great pointer. We could probably use `TensorMap` on our existing data structures.
